web/packages/teleterm/src/ui/services/modals/modalsService.ts

```diff
@@ -89,21 +122,42 @@ export class ModalsService extends ImmutableStore<State> {
    * Dialogs are displayed in the order they arrive, with the most recent one on top.
    * This allows actions that need further steps to be completed.
    *
-   * The returned closeDialog function can be used to close the dialog and automatically call the
-   * dialog's onCancel callback (if present).
+   * The passed `abortSignal` or the returned `closeDialog` function can be used to close the dialog
+   * and automatically call the dialog's onCancel callback (if present).
    */
-  openImportantDialog(dialog: Dialog): { closeDialog: () => void; id: string } {
+  openImportantDialog(
+    dialog: Dialog,
+    abortSignal?: AbortSignal
```

Comment:
Do we need to support signals in `openImportantDialog`? Shouldn't important dialogs stay open no matter what? Unlike regular dialogs, they can be stacked on each other, while there can be only a single regular dialog open at a time.

Reply:
At first I planned to add it only to `openRegularDialog`. However, I realized that when launching a deep link I may want to close all dialogs, not only the regular ones, because of the following:
>1. Activating a deep link may require changing the workspace, and we don't want to see dialogs from the previous one.
>2. A login dialog could be covered by an important dialog.
>3. The user could be confused, since Connect My Computer or Authorize Web Session documents would be displayed below a dialog.

But if you mean that we don't need to accept signals as an argument in `openImportantDialog`, I did it mainly to make the API of these two methods the same. But yes, we don't use it anywhere, so I can remove it.

Reply:
> At first I planned to add it only to `openRegularDialog`. However, I realized that when launching a deep link I may want to close all dialogs, not only the regular ones, because of the following (…)

On one hand it makes sense, on the other hand even if you close all dialogs, there might be a goroutine waiting for the mutex to get unlocked and then immediately create an important dialog soon after you close the existing ones.

From the perspective of the app as a whole, I think it makes sense to keep invariants simple, e.g., "Important modals are always shown on top and never get closed without user interaction" vs "Important modals are always shown on top, but in certain situations they might get closed so that we can show a regular modal".

---

Ultimately, in a case where there's both a deep link launch and an important modal to show, we have to decide what we deem more "important" to present to the user. I'd argue that for deep links, there's a clear understanding on the side of the user that clicking this link is going to launch Teleport Connect. Important modals on the other hand are caused by 3rd-party apps where the user might not necessarily make the connection that connecting to a db through DBeaver is going to show something in Connect.

---

On the third hand, we can imagine such situation where the user connects through DBeaver and gets a relogin modal for workspace A. They ignore it, switch to a browser and then open a deep link to workspace B. As you mentioned, what sense would it make to show the user the relogin modal over whatever the deep link wanted to show?

So I'm inclined to agree here, but then what's the purpose of important modals? I guess we could sketch out a hierarchy where it's basically:

deep links > important modals > regular modals

From that perspective, this behavior where we close all modals makes sense I suppose. Let me know what you think.

Reply:
>On one hand it makes sense, on the other hand even if you close all dialogs, there might be a goroutine waiting for the mutex to get unlocked and then immediately create an important dialog soon after you close the existing ones.

Yeah, that can happen, but I think we have to live with this possibility. 

>So I'm inclined to agree here, but then what's the purpose of important modals? I guess we could sketch out a hierarchy where it's basically:
>
>deep links > important modals > regular modals
>
>From that perspective, this behavior where we close all modals makes sense I suppose. Let me know what you >think.

The other way to think about it is that one action interrupts another and the last one always wins.
As we assumed, opening a deep link and showing modals are actions that can happen at the same time, so we have two choices:
* let the user complete the modal and then continue in the document,
* close the modal for them and go straight to the document.

I believe it makes more sense to interrupt the previous action, ant let the user perform the most recent one.
I'm not even sure we can talk about hierarchy here (deep links > important modals > regular modals), because an important modal can interrupt the action in a document opened by a deep link. We just don't have to cancel anything, the app structure forces you to complete the modal first.

Reply:
That's a useful way of framing this. I suppose the biggest difference is that important modals can interrupt an action and then lets you pick it up – an important modal might not originate from the user directly, so we don't want to lose anything. A deep link discards the current action and does not let you pick it up, but in this case the deep link was launched by the user, so we can be more confident that they indented to cancel the current action.